ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,363 @@
+  /**
+   * Used to describe the state of the LogSegment.
+   */
+  enum State {
+    FREE,
+    ACTIVE,
+    SEALED
+  }
+
+  private final FileChannel fileChannel;
+  private final File file;
+  private final long capacityInBytes;
+  private final String name;
+  private final Pair<File, FileChannel> segmentView;
+  private final StoreMetrics metrics;
+  private final AtomicLong endOffset;
+  private final AtomicLong refCount = new AtomicLong(0);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  /**
+   * The state of the segment.
+   * <p/>
+   * A LogSegment once instantiated can only go from FREE -> ACTIVE -> SEALED in the natural course of operations, never
+   * backward. However it can be externally manipulated to move in any direction.
+   */
+  State state;
+  /**
+   * Reference to the segment that logically "follows" this segment.
+   */
+  LogSegment next = null;
+
+  /**
+   * Creates a LogSegment abstraction.
+   * @param name the desired name of the segment.
+   * @param file the backing file for this segment.
+   * @param capacityInBytes the intended capacity of the segment
+   * @param metrics the {@link StoreMetrics} instance to use.
+   * @throws IOException if the file cannot be read or created
+   */
+  LogSegment(String name, File file, long capacityInBytes, StoreMetrics metrics)
+      throws IOException {
+    if (!file.exists() || !file.isFile()) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + " does not exist or is not a file");
+    }
+    this.file = file;
+    this.name = name;
+    this.capacityInBytes = capacityInBytes;
+    this.metrics = metrics;
+    fileChannel = Utils.openChannel(file, true);
+    segmentView = new Pair<>(file, fileChannel);
+    // state is always initialized to FREE and the end offset is always initialized to 0.
+    // externals will set the correct values for these two variables
+    state = State.FREE;
+    endOffset = new AtomicLong(0);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Guarantees that the {@code buffer} is either written in its entirety in this segment or is transparently forwarded
+   * to the next segment (if one exists). A write is not started if it cannot be completed.
+   * <p/>
+   * Can cause state change from {@link State#FREE} -> {@link State#ACTIVE} if this is the first write to the segment.
+   * Can cause state change from {@link State#ACTIVE} -> {@link State#SEALED} if this write cannot be accommodated in
+   * the segment.
+   * @param buffer The buffer from which data needs to be written from
+   * @return the number of bytes written.
+   * @throws IllegalArgumentException if there is not enough space for {@code buffer} and forwarding is not possible
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space
+   */
+  @Override
+  public int appendFrom(ByteBuffer buffer)
```

Comment:
Yes and those locks will continue. The atomicity required is both the log and index together for most operations (like store put and delete) - it will have to done under a lock anyway. 

Gets still don't have to take a lock and compaction is designed to avoid locking. Trying to make this thread safe only complicates the code without any extra benefits is what I think.